lib/logic/player/content.dart

```diff
@@ -23,137 +22,74 @@ extension QuickActionSerialization on QuickAction {
 class ContentMap<V> {
-  /// Creates a content map from initial value [map].
-  ///
-  /// If none specified, will initialize with empty map.
-  ContentMap([Map<Type, V>? map]) : _map = map ?? {};
-
-  final Map<Type, V> _map;
-
-  /// Map values.
-  Iterable<V> get values => _map.values;
-
-  /// Map entries.
-  Iterable<MapEntry<Type, V>> get entries => _map.entries;
-
-  /// Returns a value per `T` [Content] from the map.
-  ///
-  /// If [key] was explicitly provided, will use it instead.
-  V? getValue<T extends Content>([Type? key]) {
-    assert(
-      Content.enumerate().contains(key ?? T),
-      "Specified type must be a subtype of Content",
+  /// The value for [ContentType.song].
+  V songValue;
+
+  /// The value for [ContentType.album].
+  V albumValue;
+
+  /// The value for [ContentType.playlist].
+  V playlistValue;
+
+  /// The value for [ContentType.artist].
+  V artistValue;
+
+  ContentMap({
+    required this.songValue,
+    required this.albumValue,
+    required this.playlistValue,
+    required this.artistValue,
+  });
+
+  /// Create a content map and initialize it with the values returned by the [factory],
+  /// which will be called once with each [ContentType].
+  factory ContentMap.fromFactory(V Function(ContentType contentType) factory) {
+    return ContentMap(
+      songValue: factory(ContentType.song),
+      albumValue: factory(ContentType.album),
+      playlistValue: factory(ContentType.playlist),
+      artistValue: factory(ContentType.artist),
     );
-    return _map[key ?? T];
   }
 
-  /// Puts a [value] typed with `T` into the map.
-  ///
-  /// If [key] was explicitly provided, will use it instead.
-  void setValue<T extends Content>(V value, {Type? key}) {
-    assert(
-      Content.enumerate().contains(key ?? T),
-      "Specified type must be a subtype of Content",
-    );
-    _map[key ?? T] = value;
-  }
+  /// Map values.
+  Iterable<V> get values => ContentType.values.map((type) => get(type));
 
-  /// Look up the value of [key], or add a new entry if it isn't there.
-  ///
-  /// If [key] was explicitly provided, will use it instead.
-  V putIfAbsent<T extends Content>(V Function() ifAbsent, {Type? key}) {
-    assert(
-      Content.enumerate().contains(key ?? T),
-      "Specified type must be a subtype of Content",
-    );
-    return _map.putIfAbsent(key ?? T, ifAbsent);
+  /// Map entries.
+  Iterable<MapEntry<ContentType, V>> get entries => ContentType.values.map((type) => MapEntry(type, get(type)));
+
+  /// Returns the value for the [type] from the map.
+  V get(ContentType type) {
+    switch (type) {
+      case ContentType.song:
+        return songValue;
+      case ContentType.album:
+        return albumValue;
+      case ContentType.playlist:
+        return playlistValue;
+      case ContentType.artist:
+        return artistValue;
+    }
   }
 
-  /// Removes all entries from the map.
-  void clear() {
```

Comment:
You could do

```dart
extension ContentMapClearExtension on ContentMap<Object?> {
  void clear() {
    songValue = null;
    albumValue = null;
    playlistValue = null;
    artistValue = null;
  }
}
```

Reply:
I would do into
```dart
void clear() {
  for (final contentType in ContentType.values) {
    set(key: contentType, null);
  }
}
```
to avoid missing a ContentType when we add one.

https://github.com/nt4f04uNd/sweyer/pull/70#discussion_r912414199 would be the only usage of this. Let me know if you still want me to add this extension if you decide against setting them to `null`.

Reply:
Also, the extension has a massive problem:

`Object?` is a parent type of everything, including non-null types.
<details>
<summary> Graphic from dart.dev </summary>

[![Dart type hierarchy](https://dart.dev/null-safety/understanding-null-safety/top-and-bottom.png)](https://dart.dev/null-safety/understanding-null-safety#top-and-bottom)
</details>

This means it matches everything, including non nullable types:
```dart
final ContentMap<int> x = ...;
final ContentMap<Object?> y = x;
x.clear();
```

This problem is not unique to `ContentMap`, I can do the same with a `List`:
```dart
final a List<int> = [1,2];
final List<Object?> b = a;
b.add(null);
```

Reply:
Yeah, you are right, I thought it would only work on nullable types.